src/Cli/dotnet/commands/dotnet-test/SolutionAndProjectUtility.cs

```diff
@@ -0,0 +1,128 @@
+                    ..Directory.GetFiles(directory, "*.slnx", SearchOption.TopDirectoryOnly)];
+
+                // If more than a single sln file is found, an error is thrown since we can't determine which one to choose.
+                if (possibleSolutionPaths.Count() > 1)
+                {
+                    VSTestTrace.SafeWriteTrace(() => string.Format(CommonLocalizableStrings.MoreThanOneSolutionInDirectory, directory));
+                    return false;
+                }
+                // If a single solution is found, use it.
+                else if (possibleSolutionPaths.Count() == 1)
+                {
+                    // Get project file paths to check if there are any projects in the directory
+                    string[] possibleProjectPaths = GetProjectFilePaths(directory);
+
+                    if (possibleProjectPaths.Count() == 0)
+                    {
+                        projectOrSolutionFilePath = possibleSolutionPaths[0];
+                        isSolution = true;
+                        return true;
+                    }
+                    else // If both solution and project files are found, return false
+                    {
+                        VSTestTrace.SafeWriteTrace(() => LocalizableStrings.CmdMultipleProjectOrSolutionFilesErrorMessage);
+                        return false;
+                    }
+                }
+                // If no solutions are found, look for a project file
+                else
+                {
+                    string[] possibleProjectPath = GetProjectFilePaths(directory);
+
+                    // No projects found throws an error that no sln nor projects were found
+                    if (possibleProjectPath.Count() == 0)
+                    {
+                        VSTestTrace.SafeWriteTrace(() => LocalizableStrings.CmdNoProjectOrSolutionFileErrorMessage);
+                        return false;
+                    }
+                    // A single project found, use it
+                    else if (possibleProjectPath.Count() == 1)
+                    {
+                        projectOrSolutionFilePath = possibleProjectPath[0];
+                        return true;
+                    }
+                    // More than one project found. Not sure which one to choose
+                    else
+                    {
+                        VSTestTrace.SafeWriteTrace(() => string.Format(CommonLocalizableStrings.MoreThanOneProjectInDirectory, directory));
+                        return false;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+
+        private static string[] GetProjectFilePaths(string directory)
+        {
+            var projectFiles = Directory.GetFiles(directory, "*.*proj", SearchOption.TopDirectoryOnly)
+                .Where(f => IsProjectFile(f))
+                .ToArray();
```

Comment:
Nit: No need to build the array on the first try as you are filtering it down after.

```suggestion
            var projectFiles = Directory.EnumerateFiles(directory, "*.*proj", SearchOption.TopDirectoryOnly)
                .Where(IsProjectFile)
                .ToArray();
```